### PR TITLE
build(ci): publish artifacts to Terasology artifactory

### PR DESCRIPTION
**Pull request by skaldarnar**
After some experimentation on my [fork ](https://github.com/skaldarnar/JNBullet) I finally got a [successful workflow run](https://github.com/skaldarnar/JNBullet/actions/runs/8308383188) that published artifacts for [1.0.4-SNAPSHOT](https://artifactory.terasology.io/ui/repos/tree/General/libs-snapshot-local/org/terasology/jnbullet/JNBullet/1.0.4-SNAPSHOT/), including the zipped native libraries for Windows, Mac, and Linux. 

I've already added the respective [Secrets ](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions) to the repository.

Contributes to https://github.com/MovingBlocks/Terasology/issues/5055.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -157,7 +156,9 @@ task zipNatives(type: Zip){
 }
 
 buildNatives.dependsOn generateSources
-zipNatives.dependsOn buildNatives
+// Building natives is a prerequisite for zipping them, but we don't want to re-compute them every time.
+// Also, we build natives on different platforms and combine them to a single zip, so we can't just depend on the native tasks anyway.
+//zipNatives.dependsOn buildNatives
```
Comment (jdrueckert) [inline build.gradle:161]: do we still need this?

File: .github/workflows/allInOne.yml [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    steps:
+      - name: Install SWIG
+        run: sudo apt-get install swig mingw-w64
+      - uses: actions/checkout@v4
+        with:
+          submodules: true
+      - name: Set up JDK
+        uses: actions/setup-java@v4
+        with:
+          java-version: '11'
```
Comment (jdrueckert) [inline .github/workflows/allInOne.yml:24]: would this also work with java 17?

File: .github/workflows/allInOne.yml [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+      - name: Upload artifacts
+        uses: actions/upload-artifact@v4
+        with:
+          name: linux-windows-amd64-artifacts
+          path: |
+            build/natives/*/*.so
+            build/natives/*/*.dll
+            build/natives/*/*.dylib
+  build-macos-amd64:
+    runs-on: macos-12
```
Comment (jdrueckert) [inline .github/workflows/allInOne.yml:39]: would this also work with mac OS 14?

File: .github/workflows/allInOne.yml [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        run: |
+          brew install swig
+          brew install cmake
+      - uses: actions/checkout@v4
+        with:
+          submodules: true
+      - name: Set up JDK
+        uses: actions/setup-java@v4
+        with:
+          java-version: '11'
```
Comment (jdrueckert) [inline .github/workflows/allInOne.yml:52]: same here

File: .github/workflows/allInOne.yml [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
```
Comment (jdrueckert) [inline .github/workflows/allInOne.yml:80]: same here

File: .github/workflows/allInOne.yml [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+      - name: Build
+        run: ./gradlew buildNatives
+      - name: Upload artifacts
+        uses: actions/upload-artifact@v4
+        with:
+          name: linux-windows-amd64-artifacts
+          path: |
+            build/natives/*/*.so
+            build/natives/*/*.dll
+            build/natives/*/*.dylib
```
Comment (jdrueckert) [inline .github/workflows/allInOne.yml:37]: arent's these like platform specific?
.so linux, .dll windows, .dylib mac?
if so, can we remove .dylib here?

File: .github/workflows/allInOne.yml [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+      - name: Build
+        run: ./gradlew native_macosx_amd64_clang
+      - name: Upload artifacts
+        uses: actions/upload-artifact@v4
+        with:
+          name: macosx-amd64-artifacts
+          path: |
+            build/natives/*/*.so
+            build/natives/*/*.dll
+            build/natives/*/*.dylib
```
Comment (jdrueckert) [inline .github/workflows/allInOne.yml:65]: same here, remove .so and .dll?

File: .github/workflows/allInOne.yml [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+      - name: Build
+        run: ./gradlew native_macosx_aarch64_clang
+      - name: Upload artifacts
+        uses: actions/upload-artifact@v4
+        with:
+          name: macosx-aarch64-artifacts
+          path: |
+              build/natives/*/*.so
+              build/natives/*/*.dll
+              build/natives/*/*.dylib
```
Comment (jdrueckert) [inline .github/workflows/allInOne.yml:93]: same here, remove .so and .dll?

File: .github/workflows/allInOne.yml [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    runs-on: ubuntu-latest
+    needs: [validate-gradle-wrapper,build-linux-windows-amd64, build-macos-amd64, build-macos-aarch64]
+    if: github.ref == 'refs/heads/master'
+    steps:
+      - name: Install SWIG
+        run: sudo apt-get install swig mingw-w64
+      - name: Set up JDK
+        uses: actions/setup-java@v4
+        with:
+          java-version: '11'
```
Comment (jdrueckert) [inline .github/workflows/allInOne.yml:104]: same here

File: .github/workflows/allInOne.yml [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+      - name: Build and Zip Natives
+        run: ./gradlew build zipNatives      
+      - run: |
+          ls -R build/natives/
+          ls -R build/libs/
+      - name: Publish
+        run: ./gradlew -Dorg.gradle.internal.publish.checksums.insecure=true publish -PmavenUser=${artifactoryUser} -PmavenPass=${artifactoryPass}
+        env:
+          artifactoryUser: ${{ secrets.ARTIFACTORY_USER }}
+          artifactoryPass: ${{ secrets.ARTIFACTORY_PASS }}
```
Comment (jdrueckert) [inline .github/workflows/allInOne.yml:125]:
```suggestion
          artifactoryPass: ${{ secrets.ARTIFACTORY_PASS }}

```

**Review by jdrueckert — APPROVED** [review]

File: build.gradle (hunk 1 above)
Comment (skaldarnar) [reply build.gradle:161]: I think that some kind of check could be helpful, and until we have that, leave this explanation in why we don't have this dependency anymore.

For instance, check whether there are _any_ files included at all, if there are none, print a warning about "Did you forget to run `./gradlew buildNatives` or copy in built natives to `build/natives/*`?"

File: .github/workflows/allInOne.yml (hunk 2 above)
Comment (skaldarnar) [reply .github/workflows/allInOne.yml:24]: Probably - the Gradle build file requires Java 1.8 still 🤷 

https://github.com/MovingBlocks/JNBullet/blob/85b1a57a8e70a8538607b097a75e79c81fba2ede/build.gradle#L26-L29

File: .github/workflows/allInOne.yml (hunk 3 above)
Comment (skaldarnar) [reply .github/workflows/allInOne.yml:39]: `macos-12` is the standard for `macos-latest` for [GitHub-hosted runners](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories) and translates to an `amd64` Intel-based Mac.

`macos-14` is currently a beta runner and translates to an M1 CPU (Apple Silicion).

Thus, the workflow is using "the latest macOS runner" (`macos-12`) and a "beta runner for M1" (`macos-14`).

I gave it a test on my M1 Mac, and it looks like `./gradlew buildNatives` is able to build both targets for Apple Silicon (aarch64) and older Intel-based Macs (amd64) on the same machine. So maybe we can use just `macos-14` here 🤔

File: .github/workflows/allInOne.yml (hunk 6 above)
Comment (skaldarnar) [reply .github/workflows/allInOne.yml:37]: Yes, with the split into the different jobs we can make these more specific 👍 

However, I started looking into using a single job with a [matrix over the OS](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs), and in that case the "Upload Artifacts" step would need this general pattern again, I think 🤔

File: .github/workflows/allInOne.yml (hunk 2 above)
Comment (jdrueckert) [reply .github/workflows/allInOne.yml:24]: okay, so we might want to update that at some point like we did for the other repos.
can be done in a follow-up PR though

File: .github/workflows/allInOne.yml (hunk 6 above)
Comment (skaldarnar) [reply .github/workflows/allInOne.yml:37]: ... but if we only need two different runners (`ubuntu-latest` for Linux + Windows and `macos-14` for macOS amd64 and aarch64) we might just keep it as separate jobs without the matrix 🤔 

A single job would show that the steps to build the natives are quite similar, and that only the setup of native dependencies is a bit different...

File: .github/workflows/allInOne.yml (hunk 3 above)
Comment (jdrueckert) [reply .github/workflows/allInOne.yml:39]: might be worth at least a try.
otherwise, maybe add a code comment with the explanation you just shared

**Review by jdrueckert — APPROVED** [review]

**Pull request merged**